### PR TITLE
Show 'user typing' indicator

### PR DESCRIPTION
**Pull request by borisyankov**
* Handle 'typing' events from server
* Add new 'typing' reducer, actions, selectors, and tests
* Detect if there is typing in current narrow + tests
* Show visual indicator if a user is typing

Adds to the end of the message:

screen shot 2017-04-10 at 10 10 14 pm

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@borisyankov, it looks like you've already signed the Dropbox CLA. Thanks!

File: src/chat/chatSelectors.js [hunk 1]
```diff
@@ -39,3 +40,20 @@ export const getRecentConversations = (state) => {
     .map(recipient => recipient[0]);
   return sortedByTimestamp;
 };
+
+export const getCurrentTypingUser = state => {
```
Comment (neerajwahi) [inline src/chat/chatSelectors.js]: What does this function do? What if there are multiple users typing in a narrow

Also please add comments

File: src/events/eventToAction.js [hunk 2]
```diff
@@ -112,6 +119,13 @@ export default (state, event) => {
         type: EVENT_UPDATE_MESSAGE_FLAGS,
       };
 
+    case 'typing':
+      return {
+        ...event,
+        selfEmail: state.accounts[0].email,
+        type: opToActionTyping[event.op],
```
Comment (neerajwahi) [inline src/events/eventToAction.js]: I really dislike this `opToAction` pattern

File: src/chat/chatSelectors.js (hunk 1 above)
Comment (borisyankov) [inline src/chat/chatSelectors.js]: It was intentionally not handling two people typing, but I'll add that.

**Comment by borisyankov** [comment]
@neerajwahi Added support for multiple people typing at the same time.

File: src/main/MainScreenContainer.js [hunk 3]
```diff
@@ -55,6 +59,7 @@ export default connect(state => ({
   caughtUp: state.chat.caughtUp,
   narrow: state.chat.narrow,
   mute: state.mute,
+  typingUser: getCurrentTypingUsers(state),
```
Comment (Sam1301) [inline src/main/MainScreenContainer.js]: is there a typo here? `typingUser` => `typingUsers`?

Comment (borisyankov) [inline src/main/MainScreenContainer.js]: Good catch. It is a bug from a rebase. Fixing right now.

Comment (borisyankov) [inline src/main/MainScreenContainer.js]: Fixed.

File: src/message/MessageTyping.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  handleAvatarPress = () => this.props.pushRoute('account-details', this.props.email);
+
+  render() {
+    const { avatarUrl, fullName } = this.props;
+
+    return (
+      <View style={styles.message}>
+        <Avatar
+          avatarUrl={avatarUrl}
```
Comment (Sam1301) [inline src/message/MessageTyping.js]: for user uploaded profile pictures, we might need to obtain the full url
`/user_avatars/2/7d16d3a020d64aaee34aa0ae2ed4.png?x=x&version=1`

File: src/message/MessageTyping.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  render() {
+    const { users } = this.props;
+    const text = `... ${users.length > 1 ? 'are' : 'is'} typing`;
+
+    return (
+      <View style={styles.message}>
+        {users.map(user => (
+          <View key={user.email} style={styles.avatar}>
+            <Avatar
+              avatarUrl={user.avatarUrl}
```
Comment (Sam1301) [inline src/message/MessageTyping.js]: for user uploaded profile pictures, we might need to obtain the full url
/user_avatars/2/7d16d3a020d64aaee34aa0ae2ed4.png?x=x&version=1

Comment (borisyankov) [inline src/message/MessageTyping.js]: Good catch. I thought we were handling this inside Avatar but I guess not.

Comment (Sam1301) [inline src/message/MessageTyping.js]: opened for it #567

File: src/chat/__tests__/chatSelectors-test.js [hunk 6]
```diff
@@ -153,6 +158,110 @@ describe('getRecentConversations', () => {
...
+      fullName: 'Mark Dark',
+    };
+
+    const state = {
+      accounts: [{ email: 'me@example.com' }],
+      chat: {
+        narrow: privateNarrow('john@example.com'),
+      },
+      typing: {
+        'john@example.com': [1, 2],
```
Comment (Sam1301) [inline src/chat/__tests__/chatSelectors-test.js]: didn't understand this test case; how can `mark@example.com` user id be added to typing state with `john@example.com` as recipient? (Assuming self email is `me@example.com`)

File: src/message/MessageList.js [hunk 7]
```diff
@@ -80,6 +82,7 @@ export default class MessageList extends React.PureComponent {
         <LoadingIndicator active={fetching.older} caughtUp={caughtUp.older} />
         {messageList}
         {!singleFetchProgress && fetching.newer && <LoadingIndicator active />}
+        {typingUsers && <MessageTyping users={typingUsers} />}
```
Comment (Sam1301) [inline src/message/MessageList.js]: typo here: pushRoute also needs to be passed

Comment (borisyankov) [inline src/message/MessageList.js]: Nice catch. A good argument for doing a `connect` via redux, but again, maybe later.

File: src/chat/__tests__/chatSelectors-test.js (hunk 6 above)
Comment (borisyankov) [inline src/chat/__tests__/chatSelectors-test.js]: Hmm, yeah, a bit of a mistake, it is not entirely correct from a human reading the test. The test itself is correct, as that part is kinda not relevant and not being tested (still is confusing and will fix it)

**Pull request merged**